plugins/article_import/lib/foodsoft_article_import/bnn.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+# Module for translation and parsing of BNN-files (www.n-bnn.de)
+#
+module FoodsoftArticleImport
+  module Bnn
+    @@codes = {} # rubocop:todo Style/ClassVars
+    @@midgard = {} # rubocop:todo Style/ClassVars
+    # Loads the codes_file config/bnn_codes.yml into the class variable @@codes
+    def self.load_codes(custom_file_path = nil)
+      @gem_lib = File.expand_path '..', __dir__
+      dir = File.join @gem_lib, 'foodsoft_article_import'
+      begin
+        @@codes = YAML.safe_load(File.open(File.join(dir, 'bnn_codes.yml'))).symbolize_keys # rubocop:todo Style/ClassVars
+        if custom_file_path
+          custom_codes = YAML.safe_load(File.open(custom_file_path)).symbolize_keys
+          custom_codes.each_key do |key|
+            custom_codes[key] = custom_codes[key].merge @@codes[key] if @@codes.keys.include?(key)
+            @@codes = @@codes.merge custom_codes # rubocop:todo Style/ClassVars
+          end
+        end
+        @@midgard = YAML.safe_load(File.open(File.join(dir, 'midgard_codes.yml'))).symbolize_keys # rubocop:todo Style/ClassVars
+      rescue StandardError => e
+        raise "Failed to load bnn_codes: #{dir}/{bnn,midgard}_codes.yml: #{e.message}"
+      end
+    end
+
+    $missing_bnn_codes = [] # rubocop:todo Style/GlobalVars
+
+    # translates codes from BNN to foodsoft-code
+    def self.translate(key, value)
+      if @@codes[key][value]
+        @@codes[key][value]
+      elsif @@midgard[key]
+        @@midgard[key][value]
+      elsif !value.nil?
+        $missing_bnn_codes << value # rubocop:todo Style/GlobalVars
+        nil
+      end
+    end
+
+    NAME = 'BNN (CSV)'
+    OUTLIST = false
+    OPTIONS = {
+      encoding: 'IBM850',
+      col_sep: ';'
+    }.freeze
+
+    # parses a bnn-file
+    def self.parse(file, custom_file_path: nil, **opts)
+      custom_file_path ||= nil
+      encoding = opts[:encoding] || OPTIONS[:encoding]
+      col_sep = opts[:col_sep] || OPTIONS[:col_sep]
+      load_codes(custom_file_path)
+      piece_unit_code = 'XPP'
+      CSV.foreach(file, { col_sep: col_sep, encoding: encoding, headers: true }).with_index(1) do |row, i|
+        # check if the line is empty
+        unless row[0] == '' || row[0].nil?
+          article = {
+            availability: %w[X V].exclude?(row[1]),
+            name: UTF8Encoder.clean(row[6]),
```

Comment:
If this was ok for now, we might leave it as it is, but:

The optional row[8] provides a cleaner, shorter article name without producer and unit. We could check if it is filled, and if not, use row[6] instead. However, there might be multiple articles with the same name in row[8], so we'd have to check for differences (producer, unit, origin, quality etc.) and specify them in brackets at the end of the article name, since articles of one supplier must have unique names in Foodsoft.
But it might also be useful to see the supplier directly in the article name.

If this is changed, `note` should be changed as well.

Reply:
In my examples, `row[8]` is unfortunately always empty.

The article name is also specified for automatic orders (see #1126). I am not sure whether the catalog name should appear in the order (the order number should of course always be the primary criterion).

Again, I see it the same way as before: When importing, information can be adjusted as desired and errors (such as rare duplicate names) can also be corrected manually. As a developer, I wouldn't like to make specifications to foodcoops how item names should be designed. I suspect that there are many different tastes.